ql/lib/codeql/kaleidoscope/Cfg.qll

```diff
@@ -0,0 +1,254 @@
+private import codeql.kaleidoscope.Ast
+private import codeql.controlflow.Cfg as CfgShared
+private import codeql.Locations
+private import codeql.kaleidoscope.controlflow.BasicBlocks
+
+private module Completion {
+  private newtype TCompletion =
+    TSimpleCompletion() or
+    TBooleanCompletion(boolean b) { b in [false, true] } or
+    TReturnCompletion()
+
+  abstract class Completion extends TCompletion {
+    abstract string toString();
+
+    predicate isValidForSpecific(AstNode e) { none() }
+
+    predicate isValidFor(AstNode e) { this.isValidForSpecific(e) }
```

Comment:
Why are both needed?

Reply:
To get around a non-monotonic recursion problem in 
```ql
class SimpleCompletion extends NormalCompletion, TSimpleCompletion {
    override predicate isValidFor(AstNode e) { not any(Completion c).isValidForSpecific(e) }
}
```